File: Recursive Descent Parser/recursiveDescentParser.cpp
#include <bits/stdc++.h>
#define MAXSTRING 1000
using namespace std;
struct node {
    vector < int> prod;
    int id;
    node () {}
    node (vector <int> _prod, int _id) {
        prod = _prod;
        id = _id;
    }
};
map <string, int> stringMap;
map <int, string> reverseMap;
vector < node > grammer[MAXSTRING];
int cnt = 0;
int ruleCnt = 0;
void clearGrammer () {
    // This function clears the grammer
    stringMap.clear();
    reverseMap.clear();
    cnt = 0;
    ruleCnt = 0;
    for (int i=0; i<MAXSTRING; i++) grammer[i].clear();
}
int convertToInt (string str) {
    if (stringMap.find(str) != stringMap.end()) return stringMap[str];
    else {
        cnt++;
        stringMap[str] = cnt;
        reverseMap[cnt] = str;
    }
    return stringMap[str];
}
vector <int> strSplit (string str) {
    string tmp = "";
    vector <int> ret;
    ret.clear();

    for (int i=0; i<str.size(); i++) {
        if (str[i] == ' ') {
            if (tmp != "") {
                ret.push_back(convertToInt(tmp));
                tmp = "";
            }
        }
        else tmp.push_back(str[i]);
    }
    if (tmp != "") ret.push_back(convertToInt(tmp));
    return ret;
}
vector <string> divideRules (string str) {
    vector <string> ret;
    string tmp = "";
    for (int i=0; i<str.size(); i++) {
        if (str[i] == '|') {
            if (tmp != "") {
                ret.push_back(tmp);
                tmp = "";
            }
        }
        else tmp.push_back(str[i]);
    }
    if (tmp != "") {
        ret.push_back(tmp);
        tmp = "";
    }
    return ret;
}
int addRule (string from, string to) {
    // This function adds a rule
    vector <string> rules;
    rules = divideRules(to);
    vector <int> prod;
    for (int i=0; i<rules.size(); i++) {
        prod = strSplit(rules[i]);
        grammer[convertToInt(from)].push_back(node(prod, ++ruleCnt));
    }
}
void printRules () {
    for (int i=1; i<cnt; i++) {
        for (int j=0; j<grammer[i].size(); j++) {
            printf("Rule %d: ", grammer[i][j].id);
            cout << reverseMap[i] << " -> " ;
            for (int k=0; k<grammer[i][j].prod.size(); k++) {
                cout << reverseMap[grammer[i][j].prod[k]] << " ";
            }
            printf("\n");
        }
    }
}
string toString (int num) {
    string ret = "";
    ret.push_back('0' + num);
    return ret;
}
int main () {
    addRule("goal", "expr");
    addRule("expr", "term expr2");
    addRule("expr2", "+ term expr2 | - term expr2 | eps");
    addRule("term", "factor term2");
    addRule("term2", "* factor term2 | / factor term2 | eps");
    addRule("factor", "number | identifier | ( expr )");
    for (int i=0; i<10; i++) addRule("number", toString(i));
    addRule("factor", "+");
    addRule("factor", "-");
    addRule("factor", "*");
    addRule("factor", "/");
    addRule("factor", "|");
    printRules();
    return 0;
}

File: Recursive Descent Parser/recursiveDescentParser.py
'''
Name: Muntasir Wahed
Roll: AE-075-007
'''


cursor = 0
savedCursor = 0
str = ''
def saveCursor():
    global cursor, savedCursor
    savedCursor = cursor
    return 1
def backtrack():
    global cursor, savedCursor
    cursor = savedCursor
    return 1
def getNextToken():
    global cursor
    if (cursor >= len(str)):
        return '$' 
    while (str[cursor] == ' '):
        cursor = cursor + 1
    nextToken = str[cursor]
    cursor = cursor + 1
    if (nextToken >= '0' and nextToken <= '9'):
        return 'num'
    elif (nextToken >= 'a' and nextToken <= 'z'):
        return 'id'
    return nextToken;
def isToken(expected):
    return int(getNextToken() == expected)


# In[93]:


# goal → expr
def goal():
    return expr()


# In[94]:


# expr → term expr2
def expr():
    return term() and expr2()


# In[95]:


# expr2 → + term expr2 | - term expr2 | ε
def expr2():
    return (saveCursor() and (isToken('+') or isToken('-')) and term() and expr2())  or (backtrack() and 1) 


# In[96]:


# term → factor term2
def term():
    return factor() and term2()


# In[97]:


# term2 → * factor term2 | / factor term2 | ε
def term2():
    return (saveCursor() and (isToken('*') or isToken('/')) and factor and term2())  or (backtrack() and 1) 


# In[98]:


# factor → number | identifier | ( expr )
def factor () :
    return (saveCursor() and (isToken('num') or isToken('id'))) or (backtrack() and isToken('(') and expr() and isToken(')'))


# In[99]:


def parse(_str):
    global str
    str = _str
    global cursor, savedCursor
    cursor = 0
    savedCursor = 0
    return (goal() and cursor == len(str))


while (True):
    str = input()
    if (str == "exit"):
        break
    else:
        print(parse(str))
